File: evd_ros_backend/evd_ros_core/src/evd_script/attribute_trace_processor.py
'''
EvDScript's AST generates callback traces, which can be hooked into at the root program.

This is just a simple subscriber for the traces to route various in-app callbacks that 
you may want. For instance, your program may require tracking changes on traces. You can
then subscribe with a type=Trace.full_type_string() and your callback. When ever a change
happens that has the type field set as specified, your supplied callback will be invoked
and the entry in the trace will be passed.

This entry is defined as the node's child changed message.
'''

class AttributeTraceProcessor:

    def __init__(self):
        self._subscribers = {}

    def subscribe_to_type(self, type, callback):
        if not type in self._subscribers.keys():
            self._subscribers[type] = []

        self._subscribers[type].append(callback)

    def unsubscribe_to_type(self, type, callback):
        if not type in self._subscribers.keys():
            return False

        if not callback in self._subscribers[type]:
            return False

        self._subscribers[type].remove(callback)
        return True

    def process_trace(self, attributeTrace):
        for entry in attributeTrace:
            if entry["type"] in self._subscribers.keys():
                for cb in self._subscribers[entry["type"]]:
                    cb(entry)

File: evd_authoring_app/src/model/evdScript/attributeTraceProcessor.js
export class AttributeTraceProcessor {

    constructor() {
        this._subscribers = {};
    }

    subscribeToType(type, callback) {
        if (! (type in this._subscribers)) {
            this._subscribers[type] = [];
        }

        this._subscribers[type].push(callback);
    }

    unsubscribeToType(type, callback) {
        if (! (type in this._subscribers)) {
            return false;
        }

        const idx = this._subscribers[type].indexOf(callback);
        if (idx >= 0) {
            this._subscribers[type].splice(idx,1);
            return true;
        } else {
            return false;
        }
    }

    processTrace(attributeTrace) {
        attributeTrace.forEach(element => {
            if (element.type in this._subscribers) {
                this.subscribers[element.type].forEach(cb => {
                    cb(element);
                });
            }
        });
    }
}